core/input/mouse.js

```diff
@@ -108,27 +116,74 @@ Mouse.prototype = {
+
+        if (this._accumulatedWheelDeltaX < 0) {
+            this._onMouseButton(pos.x, pos.y, 1, 1 << 5);
+            this._onMouseButton(pos.x, pos.y, 0, 1 << 5);
+        } else if (this._accumulatedWheelDeltaX > 0) {
+            this._onMouseButton(pos.x, pos.y, 1, 1 << 6);
+            this._onMouseButton(pos.x, pos.y, 0, 1 << 6);
+        }
+
+        this._accumulatedWheelDeltaX = 0;
+    },
+
+    _generateWheelStepY: function (pos) {
+
+        if (this._accumulatedWheelDeltaY < 0) {
+            this._onMouseButton(pos.x, pos.y, 1, 1 << 3);
+            this._onMouseButton(pos.x, pos.y, 0, 1 << 3);
+        } else if (this._accumulatedWheelDeltaY > 0) {
+            this._onMouseButton(pos.x, pos.y, 1, 1 << 4);
+            this._onMouseButton(pos.x, pos.y, 0, 1 << 4);
+        }
+
+        this._accumulatedWheelDeltaY = 0;
+    },
+
     _handleMouseWheel: function (e) {
-        if (!this._focused) { return; }
+        if (!this._focused || !this._onMouseButton) { return; }
+
+        window.clearTimeout(this._wheelStepXTimer);
+        window.clearTimeout(this._wheelStepYTimer);
 
         var pos = this._getMousePosition(e);
 
-        if (this._onMouseButton) {
-            if (e.deltaX < 0) {
-                this._onMouseButton(pos.x, pos.y, 1, 1 << 5);
-                this._onMouseButton(pos.x, pos.y, 0, 1 << 5);
-            } else if (e.deltaX > 0) {
-                this._onMouseButton(pos.x, pos.y, 1, 1 << 6);
-                this._onMouseButton(pos.x, pos.y, 0, 1 << 6);
-            }
+        var dX = e.deltaX;
+        var dY = e.deltaY;
+
+        // Pixel units unless it's non-zero.
+        // Note that if deltamode is line or page won't matter since we aren't
+        // sending the mouse wheel delta to the server anyway.
+        // The difference between pixel and line can be important however since
+        // we have a threshold that can be smaller than the line height.
+        if (e.deltaMode) {
+            dX *= this.WHEEL_LINE_HEIGHT;
+            dY *= this.WHEEL_LINE_HEIGHT;
+        }
 
-            if (e.deltaY < 0) {
-                this._onMouseButton(pos.x, pos.y, 1, 1 << 3);
-                this._onMouseButton(pos.x, pos.y, 0, 1 << 3);
-            } else if (e.deltaY > 0) {
-                this._onMouseButton(pos.x, pos.y, 1, 1 << 4);
-                this._onMouseButton(pos.x, pos.y, 0, 1 << 4);
-            }
+        this._accumulatedWheelDeltaX += dX;
+        this._accumulatedWheelDeltaY += dY;
+
+        // Generate a mouse wheel step event when the accumulated delta
+        // for one of the axes is large enough.
+        // Small delta events that do not pass the threshold get sent
+        // after a timeout.
+        if (Math.abs(this._accumulatedWheelDeltaX) > this.WHEEL_STEP) {
+            this._generateWheelStepX(pos);
+        } else {
+            this._wheelStepXTimer =
+                window.setTimeout(this._generateWheelStepX.bind(this, pos), 50);
```

Comment:
Since we already have constants, then perhaps this 50ms should also be a constant? Keeps it in sync for the two calls if nothing else.

Reply:
+1